spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java

```diff
@@ -1005,45 +1074,50 @@ public Channel createChannel(boolean transactional) {
 		public void close() {
 			if (CachingConnectionFactory.this.cacheMode == CacheMode.CONNECTION) {
 				synchronized (CachingConnectionFactory.this.connectionMonitor) {
-					if (!this.target.isOpen()
-							|| CachingConnectionFactory.this.idleConnections.size() >=
-									CachingConnectionFactory.this.connectionCacheSize) {
-						if (logger.isDebugEnabled()) {
-							logger.debug("Completely closing connection '" + this + "'");
-						}
-						if (this.target.isOpen()) {
-							RabbitUtils.closeConnection(this.target);
-						}
-						this.notifyCloseIfNecessary();
-						CachingConnectionFactory.this.openConnections.remove(this);
-						CachingConnectionFactory.this.openConnectionNonTransactionalChannels.remove(this);
-						CachingConnectionFactory.this.openConnectionTransactionalChannels.remove(this);
-					}
-					else {
-						if (!CachingConnectionFactory.this.idleConnections.contains(this)) {
+					/*
+					 * Only connectionCacheSize open idle connections are allowed.
+					 */
+					if (!CachingConnectionFactory.this.idleConnections.contains(this)) {
+						if (!this.target.isOpen()
+								|| countOpenIdleConnections() >= CachingConnectionFactory.this.connectionCacheSize) {
 							if (logger.isDebugEnabled()) {
-								logger.debug("Returning connection '" + this + "' to cache");
-							}
-							CachingConnectionFactory.this.idleConnections.add(this);
-							if (CachingConnectionFactory.this.connectionHighWaterMark
-									.get() < CachingConnectionFactory.this.idleConnections.size()) {
-								CachingConnectionFactory.this.connectionHighWaterMark
-										.set(CachingConnectionFactory.this.idleConnections.size());
+								logger.debug("Completely closing connection '" + this + "'");
 							}
+							destroy();
```

Comment:
I would like to hear back from you on the matter.

For me it is just only blocker to merge.

Thanks.